### PR TITLE
Add gas limit to Operation in contracts

### PR DESCRIPTION
**Pull request by jzaki**
## What is this PR doing?
Adding a gas limit parameter to operations (protects against aggregator griefing attacks and under-payments)

## Does this PR resolve or contribute to any issues?
#446 

## Checklist
- [x] I have manually tested these changes
- [x] Post a link to the PR in the group chat

## Guidelines
- If your PR is not ready, mark it as a draft
- The `resolve conversation` button is for reviewers, not authors
  - (But add a 'done' comment or similar)

### CODE REVIEW

File: contracts/clients/src/BlsWalletWrapper.ts [hunk 1]
```diff
@@ -221,6 +225,18 @@ export default class BlsWalletWrapper {
...
+  async opWithGasEstimate(operation: Operation): Promise<Operation> {
+    const bundle = this.sign(operation);
+    operation.gas = this.blockGasLimit;
+    const gatewayAddress = await this.walletContract.trustedBLSGateway();
+    // const gatewaySigner = new VoidSigner(gatewayAddress);
+    operation.gas = await this.walletContract
+      .connect(gatewayAddress)
+      .estimateGas.performOperation(bundle.operations[0]);
+    return operation;
+  }
```
Comment (voltrevo) [inline contracts/clients/src/BlsWalletWrapper.ts]: If you tweak this like so you can avoid the need to write `gas: 0` and also avoid mutating the operation passed in, both of which are rather confusing.

```suggestion
  async opWithGasEstimate(operation: Omit<Operation, "gas">): Promise<Operation> {
    operation = { ...operation, gas: this.blockGasLimit };
    const bundle = this.sign(operation);
    const gatewayAddress = await this.walletContract.trustedBLSGateway();
    // const gatewaySigner = new VoidSigner(gatewayAddress);
    operation.gas = await this.walletContract
      .connect(gatewayAddress)
      .estimateGas.performOperation(bundle.operations[0]);
    return operation;
  }
```

File: contracts/shared/helpers/TokenHelper.ts [hunk 2]
```diff
@@ -71,19 +71,22 @@ export default class TokenHelper {
...
-              "transfer",
-              [recipient, amount.toHexString()],
-            ),
-          },
-        ],
-      }),
+      sender.sign(
+        await sender.opWithGasEstimate({
+          nonce,
+          gas: 0,
```
Comment (voltrevo) [inline contracts/shared/helpers/TokenHelper.ts]:
```suggestion
```

File: contracts/shared/helpers/callProxyAdmin.ts [hunk 3]
```diff
@@ -41,9 +41,12 @@ export async function proxyAdminBundle(
...
     encodedFunction: encodedWalletAdminCall,
   };
-  const bundle: Bundle = wallet.sign({
-    nonce: await wallet.Nonce(),
-    actions: [action],
-  });
+  const bundle: Bundle = wallet.sign(
+    await wallet.opWithGasEstimate({
+      nonce: await wallet.Nonce(),
+      gas: 0,
```
Comment (voltrevo) [inline contracts/shared/helpers/callProxyAdmin.ts]:
```suggestion
```

File: contracts/test/walletAction-test.ts [hunk 4]
```diff
@@ -114,16 +114,19 @@ describe("WalletActions", async function () {
...
-          ethValue: ethToTransfer,
-          contractAddress: recvWallet.walletContract.address,
-          encodedFunction: "0x",
-        },
-      ],
-    });
+    const tx = sendWallet.sign(
+      await sendWallet.opWithGasEstimate({
+        nonce: await sendWallet.Nonce(),
+        gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/test/walletAction-test.ts [hunk 5]
```diff
@@ -237,89 +247,98 @@ describe("WalletActions", async function () {
     await (
       await fx.verificationGateway.processBundle(
         fx.blsWalletSigner.aggregate([
-          sender1.sign({
-            nonce: await sender1.Nonce(),
+          sender1.sign(
+            await sender1.opWithGasEstimate({
+              nonce: await sender1.Nonce(),
+              gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/test/walletAction-test.ts [hunk 6]
```diff
@@ -237,89 +247,98 @@ describe("WalletActions", async function () {
...
+                    ],
+                  ),
+                },
+              ],
+            }),
+          ),
+          sender2.sign(
+            await sender2.opWithGasEstimate({
+              nonce: await sender2.Nonce(),
+              gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/test/walletAction-test.ts [hunk 7]
```diff
@@ -237,89 +247,98 @@ describe("WalletActions", async function () {
...
+  it("should prevent other actions within an operation when one fails", async () => {
+    const th = new TokenHelper(fx);
+    const [sender, recipient] = await th.walletTokenSetup();
+
+    const r: ContractReceipt = await (
+      await fx.verificationGateway.processBundle(
+        sender.sign(
+          await sender.opWithGasEstimate({
+            nonce: await sender.Nonce(),
+            gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/clients/src/BlsWalletWrapper.ts (hunk 1 above)
Comment (voltrevo) [reply contracts/clients/src/BlsWalletWrapper.ts]: Also, I'm not sure there's much of a use-case for `opWithGasEstimate` without also immediately passing the result to `sign`, so perhaps there should be `signWithGasEstimate` that does both?

File: contracts/test/walletAction-test.ts [hunk 8]
```diff
@@ -160,17 +163,20 @@ describe("WalletActions", async function () {
...
-              contractAddress: mockAuction.address,
-              encodedFunction:
-                mockAuction.interface.encodeFunctionData("buyItNow"),
-            },
-          ],
-        }),
+        sendWallet.sign(
+          await sendWallet.opWithGasEstimate({
+            nonce: BigNumber.from(1),
+            gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/test/walletAction-test.ts [hunk 9]
```diff
@@ -183,19 +189,23 @@ describe("WalletActions", async function () {
...
-            "walletFromHash",
-            [fx.blsWalletSigner.getPublicKeyHash(wallet.privateKey)],
-          ),
-        },
-      ],
-    });
+    const tx = wallet.sign(
+      await wallet.opWithGasEstimate({
+        nonce: await wallet.Nonce(),
+        gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/test/walletAction-test.ts [hunk 10]
```diff
@@ -351,22 +370,26 @@ describe("WalletActions", async function () {
...
-        encodedFunction: testToken.interface.encodeFunctionData("transfer", [
-          recvWallet.address,
-          th.userStartAmount.toString(),
-        ]),
-      })),
-    });
+    const tx = wallets[0].sign(
+      await wallets[0].opWithGasEstimate({
+        nonce,
+        gas: 0,
```
Comment (voltrevo) [inline contracts/test/walletAction-test.ts]:
```suggestion
```

File: contracts/clients/src/BlsSigner.ts [hunk 11]
```diff
@@ -80,7 +80,9 @@ export default class BlsSigner extends Signer {
       this.provider,
     );
 
-    const bundle = this.wallet.sign({ nonce, actions: [action] });
+    const bundle = this.wallet.sign(
+      await this.wallet.opWithGasEstimate({ nonce, gas: 0, actions: [action] }),
```
Comment (voltrevo) [inline contracts/clients/src/BlsSigner.ts]:
```suggestion
      await this.wallet.opWithGasEstimate({ nonce, actions: [action] }),
```

File: contracts/clients/src/BlsSigner.ts [hunk 12]
```diff
@@ -177,7 +179,9 @@ export default class BlsSigner extends Signer {
...
```
Comment (voltrevo) [inline contracts/clients/src/BlsSigner.ts]:
```suggestion
      await this.wallet.opWithGasEstimate({ nonce, actions: [action] }),
```

**Review by voltrevo — DISMISSED** [review]
I think we should change `opWithGasEstimate` a bit (see details in the comment next to it), but otherwise I think this is ok.

The test failures should be due to not updating the `bls-wallet-clients` module. [See `yarn publish-experimental` and its surrounding docs](https://github.com/web3well/bls-wallet/blob/main/docs/local_development.md#aggregator-1).

Also, this PR should be targeting `contract-updates` instead of `main`.

File: contracts/clients/src/BlsWalletWrapper.ts (hunk 1 above)
Comment (jzaki) [reply contracts/clients/src/BlsWalletWrapper.ts]: I did the `omit` + `...` initially, but I think `signWithGasEstimate` will be clearer.

**Comment by jzaki** [comment]
> 

Had targeted main initially hoping the local client + agg updates+tests would translate to CI success, and not need the intermediate branch.

**Pull request closed without merging**